chip_data_model.gni: add option for custom-implemented clusters

plan44 opened the pull request:
Add `zap_clusters_with_custom_implementation` option to `chip_data_model()` template, which allows to list clusters that need to be available in a device, but don't use the standard implementation in `src/app/clusters/<clustername>/<clustername>.cpp`.

# Usage
To exclude a particular cluster's standard implementation in favor of an application level implementation:
- in the BUILD.gn file for the application specific `.zap` file
- in the instantiation of the `chip_data_model` template
- add a variable `zap_clusters_with_custom_implementation`
- assign it a list of strings with cluster names, corresponding to directories in the `src/app/clusters` directory.
- this causes the cluster implementation file(s) not to be included in the build, allowing re-implementation at the application level.

Example, excluding standard implementation of level-control:
```
chip_data_model("zap") {
  zap_file = "myproject.zap"
  zap_clusters_with_custom_implementation = [ "level-control" ]
  zap_pregenerated_dir = "//zap-generated"
}
```

# Background
Some clusters, for example Level Control, are implemented with the assumption of direct low-level access to the device hardware. For bridged devices, the actual interface might be much more high level, as in my case where I have lights which are fully capable of performing parametrized smooth transitions but are NOT capable of receiving output changes in millisecond intervals. In such cases, a cluster might need application-specific re-implementation.

This changeset allows, from the application's ZAP BUILD.gn, to exclude the standard cluster implementation file, allowing re-implementation of the needed ember callbacks in a application specific file outside the connectedhomeip repo.

#### Testing
* tested to not affect normal builds not using this feature
* tested with an application which re-implements level-control in a separate .cpp file. Without the feature, linking has duplicate symbol errors (from both custom and standard implementations) - using the feature as shown in the example above, the app links and runs ok.

andy31415 reviewed (dismissed):
Marking changes requested to wait until after 1.0 branch

plan44 commented:
I think this is an important feature. Its further review was postponed to post-1.0, which would be now ;-)

andy31415 commented:
> I still think this is an important feature. Its further review was postponed to post-1.0, which would be now ;-)

The PR is marked as draft - should this not be draft anymore?
It probably also suffers from being very far down the open PR list ... less likely to get reviews.

andy31415 commented:
@plan44 - please remove this from draft if that is appropriate - that should get the bot start asking for reviews.

plan44 commented on src/app/chip_data_model.gni:
```diff
@@ -99,8 +100,23 @@ template("chip_data_model") {
 
+    _cust_impl_clusters = []
+    if (defined(invoker.zap_clusters_with_custom_implementation)) {
+      _cust_impl_clusters =  invoker.zap_clusters_with_custom_implementation
```
Yes, this works as is, without having `zap_clusters_with_custom_implementation` (tests) and also when actually using the feature (my bridge, having custom level-control implementation).

I'm a total gn newbie and if I remember correctly I gleaned how to implement that from somewhere else. Maybe these local vars don't have the re-assignment restriction?

plan44 replied:
I just rebased the PR onto master, re-ran test locally (check target) successfully without gn complaining, so I removed the draft status now. Hope this is ok.

plan44 commented:
> @plan44 - please remove this from draft if that is appropriate - that should get the bot start asking for reviews.

I will, as soon as I find time to rebase and re-test on master. Thanks for having had a look already :-)

plan44 commented:
@andy31415 I've rebased and retested a few days ago, and removed the draft status. Checks look fine except for pullaprove/license/cla, which I do not understand, as I did agree to the CLA a while ago and some PRs of mine are merged already. I have no idea how to re-agree to the CLA, if that's what the status is asking for.

bzbarsky-apple commented:
@plan44, @andy31415 is on vacation right now, as is @woody-apple, and they're the obvious people to be able to override the CLA bits....

In practice you probably want to ping people after Jan 3.

plan44 commented:
@bzbarsky-apple thanks for the info - no problem waiting until next year. Or rather than waiting, maybe finding time to prepare the libev and gaps-in-endpointId PRs ;-)

bzbarsky-apple commented on src/app/chip_data_model.gni:
```diff
@@ -137,8 +138,22 @@ template("chip_data_model") {
                                      [ invoker.zap_file ])
     }
 
+    _cust_impl_clusters = []
```
```suggestion
    _custom_impl_clusters = []
```

plan44 replied:
agreed, `cust` is too much abbreviated, `custom` is better.
I just changed all occurrences and updated the PR

bzbarsky-apple commented on src/app/chip_data_model.gni:
```diff
@@ -137,8 +138,22 @@ template("chip_data_model") {
+
     foreach(cluster, _cluster_sources) {
-      if (cluster == "door-lock-server") {
+      _cust_impl = false
```
```suggestion
      _custom_impl = false
```
